### PR TITLE
[Tunnel] implement route - IBC

### PR DESCRIPTION
**Pull request by satawatnack**
## Implementation details
- implement IBC route for Tunnel

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: x/tunnel/ibc_module.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	return nil
+}
+
+// OnRecvPacket implements the IBCModule interface
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	err := errorsmod.Wrapf(icatypes.ErrInvalidChannelFlow, "cannot receive packet on controller chain")
```
Comment (RogerKSI) [inline x/tunnel/ibc_module.go:157]: why ica and controller chain?

File: x/tunnel/ibc_module.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+) error {
+	if order != channeltypes.UNORDERED {
+		return errorsmod.Wrapf(
+			channeltypes.ErrInvalidChannelOrdering,
+			"expected %s channel, got %s",
+			channeltypes.UNORDERED,
+			order,
+		)
+	}
+
```
Comment (RogerKSI) [inline x/tunnel/ibc_module.go:199]: TBD: should it be order?

File: x/tunnel/types/tunnel.go [hunk 3]
```diff
@@ -90,6 +90,21 @@ func (t Tunnel) GetSignalIDs() []string {
...
+}
+
+// NewIBCPacketContent creates a new IBCPacketContent instance.
+func NewIBCPacketContent(channelID string, sequence uint64) *IBCPacketContent {
+	return &IBCPacketContent{
+		ChannelID: channelID,
+		Sequence:  sequence,
+	}
+}
+
```
Comment (RogerKSI) [inline x/tunnel/types/tunnel.go:107]: why is it here? think it has ibc_route file?

File: x/tunnel/types/ibc_packet_result.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	feedstypes "github.com/bandprotocol/chain/v3/x/feeds/types"
+)
+
+// IBCPacket defines the packet sent over the IBC channel
+func NewIBCPacketResult(
```
Comment (RogerKSI) [inline x/tunnel/types/ibc_packet_result.go:10]: comment start with function name.

File: proto/band/tunnel/v1beta1/tunnel.proto [hunk 5]
```diff
@@ -137,3 +145,39 @@ message TSSPacketContent {
...
+
+  // ENCODER_UNSPECIFIED is an unspecified encoder mode.
+  ENCODER_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "ENCODER_UNSPECIFIED"];
+
+  // ENCODER_FIXED_POINT_ABI is a fixed-point price abi encoder (price * 10^9).
+  ENCODER_FIXED_POINT_ABI = 1 [(gogoproto.enumvalue_customname) = "ENCODER_FIXED_POINT_ABI"];
+
+  // ENCODER_TICK_ABI is a tick abi encoder.
+  ENCODER_TICK_ABI = 2 [(gogoproto.enumvalue_customname) = "ENCODER_TICK_ABI"];
+}
```
Comment (RogerKSI) [inline proto/band/tunnel/v1beta1/tunnel.proto:183]: not use anymore?

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: x/tunnel/ibc_module.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	errorsmod "cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/bandprotocol/chain/v3/x/tunnel/keeper"
+	"github.com/bandprotocol/chain/v3/x/tunnel/types"
+)
+
```
Comment (taobun) [inline x/tunnel/ibc_module.go:20]: Please add type assertion here
```
var (
	_ porttypes.IBCModule             = (*IBCModule)(nil)
	_ porttypes.PacketDataUnmarshaler = (*IBCModule)(nil)
	_ porttypes.UpgradableModule      = (*IBCModule)(nil)
)
```

File: x/tunnel/ibc_module.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+// validateChannelParams validates the parameters of a newly created tunnel channel.
+// A valid tunnel channel must be ORDERED, use the correct port (default is 'tunnel').
+func validateChannelParams(
+	ctx sdk.Context,
+	keeper keeper.Keeper,
+	order channeltypes.Order,
+	portID string,
+) error {
+	if order != channeltypes.ORDERED {
```
Comment (taobun) [inline x/tunnel/ibc_module.go:192]: Should we add this check https://github.com/cosmos/ibc-go/blob/v8.5.2/modules/apps/transfer/ibc_module.go#L52?

File: x/tunnel/types/route_ibc.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	feedstypes "github.com/bandprotocol/chain/v3/x/feeds/types"
+)
+
+// IBCRoute defines the IBC route for the tunnel module
+var _ RouteI = &IBCRoute{}
+
+// IsChannelIDFormat checks if a channelID is in the format required on the SDK for
```
Comment (taobun) [inline x/tunnel/types/route_ibc.go:15]: Why not just import from ibc?

File: x/tunnel/types/route_ibc.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+func (r *IBCRoute) ValidateBasic() error {
+	// Validate the ChannelID format
+	if !IsChannelIDFormat(r.ChannelID) {
+		return fmt.Errorf("channel identifier is not in the format: `channel-{N}`")
+	}
+	return nil
+}
+
+// NewIBCPacketReceipt creates a new IBCPacketReceipt instance.
+func NewIBCPacketReceipt(channelID string, sequence uint64) *IBCPacketReceipt {
```
Comment (taobun) [inline x/tunnel/types/route_ibc.go:36]: Why need channel ID as param?

File: x/tunnel/ibc_module.go (hunk 7 above)
Comment (satawatnack) [reply x/tunnel/ibc_module.go:283]: added

**Review by taobun — APPROVED** [review]

**Review by RogerKSI — APPROVED** [review]

**Review by RogerKSI — APPROVED** [review]

**Pull request merged**